kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/slot-updates.k

```diff
@@ -0,0 +1,144 @@
+    rule [is-mask-true]:
+      #isMask(X:Int) => maxUInt256 ==Int X |Int ( 2 ^Int ( #getMaskShiftBits(X) +Int #getMaskWidthBits(X) ) -Int 1 )
+        requires 0 <=Int #getMaskShiftBytes(X) andBool 0 <=Int #getMaskWidthBytes(X)
+        [concrete(X), simplification, preserves-definedness]
+
+    // and is not a mask otherwise
+    rule [is-mask-false]:
+      #isMask(X:Int) => false
+        requires notBool ( 0 <=Int #getMaskShiftBytes(X) andBool 0 <=Int #getMaskWidthBytes(X) )
+        [concrete(X), simplification, preserves-definedness]
+
+    // ###########################################################################
+    // Masking lemmas
+    // ###########################################################################
+
+    // Slot updates are of the general form (SHIFT *Int VALUE) |Int (MASK &Int #asWord( SLOT:Bytes )),
+    // where the masking clears the part of the slot into which VALUE will be stored,
+    // and for VALUE to be stored correctly it first has to be shifted appropriately.
+    // Note that SHIFT and MASK are always concrete.
+    //
+    // We perform this update in several stages:
+    // 1. First, we simplify MASK &Int #asWord( SLOT ), which results in
+    //    ( VALUE *Int SHIFT ) |Int #asWord ( B1 +Bytes ... +Bytes BN ).
+    // 2. Then, we isolate the +Bytes-junct(s) that will be overwritten.
+    // 3. Then, we write the VALUE, possibly splitting the identified +Bytes-junct(s).
+    //
+    // Note that we require additional simplifications to account for the fact that
+    // VALUE and SLOT can also be concrete. In the former case, we need to extract the
+    // SHIFT appropriate, and in the latter case, the slot will appear on the LHS of the |Int.
+
+    // 1. Slot masking using &Int
+    rule [mask-b-and]:
+      MASK:Int &Int SLOT:Int =>
+        #asWord ( #buf ( 32, SLOT ) [ 32 -Int ( #getMaskShiftBytes(MASK) +Int #getMaskWidthBytes(MASK) ) := #buf ( #getMaskWidthBytes(MASK), 0 ) ] )
+        requires #rangeUInt(256, MASK) andBool #rangeUInt(256, SLOT)
+        andBool #isMask(MASK)
+        [simplification, concrete(MASK), preserves-definedness]
+
+    // 2a. |Int and +Bytes, update to be done in left
+    rule [bor-update-to-left]:
+      A |Int #asWord ( B1 +Bytes B2 ) =>
+        #asWord ( #buf ( 32 -Int lengthBytes(B2), (A /Int (2 ^Int (8 *Int lengthBytes(B2)))) |Int #asWord ( #buf (32 -Int (lengthBytes(B1) +Int lengthBytes(B2)), 0) +Bytes B1 ) ) +Bytes B2 )
+        requires #rangeUInt(256, A) andBool A modInt (2 ^Int (8 *Int lengthBytes(B2))) ==Int 0 andBool lengthBytes(B1 +Bytes B2) <=Int 32
+        [simplification, preserves-definedness]
+
+    // 2b. |Int of +Bytes, update to be done in right
+    rule [bor-update-to-right]:
+      A |Int #asWord ( B1 +Bytes B2 ) =>
+        #asWord ( B1 +Bytes #buf ( lengthBytes(B2), A |Int #asWord ( B2 ) ) )
+        requires 0 <=Int A andBool A <Int 2 ^Int (8 *Int lengthBytes(B2))
+        [simplification, preserves-definedness]
```

Comment:
We still may need restriction like `lengthBytes(B1 +Bytes B2) <=Int 32`, since `#asWord` puts a chop around the result. So say that `lengthBytes(B2) >Int 32`, and `A >Int pow256`, then the original expression `A |Int #asWord(B1 +Bytes B2)` would not be equal to the new expression, because in the new expression the upper bits of `A` would be truncated.

Reply:
https://github.com/runtimeverification/evm-semantics/blob/b3198a80e462a13423cf64cf7d5a7f5e347415b4/kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md?plain=1#L336

Reply:
We can certainly ask for `lengthBytes(B2) <= 32`, and that will also make `A <Int pow256`. Then, the chopping would chop both LHS and RHS equally, am I right?

Reply:
`lengthBytes(B2) <=Int 32` added.